Use SPIRE to rotate webhook certificates

faisal-memon opened the pull request:
**Description of change**
One of our use cases for SPIRE is managing certificates for Kubernetes webhooks.  One of the requirements there is rotating the CABundle in the ValidatingWebhookConfiguration and MutatingWebhookConfiguration.

This PR extends the existing k8s notifier to update the CABundle in webhooks. It works by a configurable label added to the xxxWebhookConfiguration, for example `spiffe.io/webhook: true`. Any xxxWebhookConfiguration with the label will have its CABundle managed by SPIRE.

This PR also adds a watcher for new xxxWebhookConfiguration's being created. For example if a MutatingWebhookConfiguration with the correct label is deployed after SPIRE is deployed, the watcher will catch the ADD event and update the CABundle.

azdagron commented:
Thanks for putting this together @faisal-memon. I think this is valuable functionality to extend the usefulness of the k8sbundle notifier. The approach seems mostly ok to me?

Other than the TODO items you mentioned, there are a few things I noticed:

-  The patches that are being created for the webhooks are created from the previous webhook object + the modified bundle config. IMO, we should scope these patches to just the bundle contents.
- The goroutine kicked off by Configure needs to be cleaned up  on subsequent calls to Configure. Alternatively, the (potentially diffferent)_webhook label needs to be passed to it so it is watching the right webhooks. We don't call Configure more than once currently, but we write Configure in a way where it can be used to reconfigure the plugin in case we allow dynamic reconfiguration in the future.

I'm happy to give a thorough review when you are ready for it.

faisal-memon commented:
> Thanks for putting this together @faisal-memon. I think this is valuable functionality to extend the usefulness of the k8sbundle notifier. The approach seems mostly ok to me?
> 
> Other than the TODO items you mentioned, there are a few things I noticed:
> 
> * The patches that are being created for the webhooks are created from the previous webhook object + the modified bundle config. IMO, we should scope these patches to just the bundle contents.
> * The goroutine kicked off by Configure needs to be cleaned up  on subsequent calls to Configure. Alternatively, the (potentially diffferent)_webhook label needs to be passed to it so it is watching the right webhooks. We don't call Configure more than once currently, but we write Configure in a way where it can be used to reconfigure the plugin in case we allow dynamic reconfiguration in the future.
> 
> I'm happy to give a thorough review when you are ready for it.

Hi Andrew,

Thanks for taking a look, I've addressed those comments and I'm ready for a more thorough review of this one. There are still some open questions I have:
- I made the changes to Configure() to make it able to be called more than once. I couldn't find a good example in the spire code of where a goroutine is being restarted in case of a configuration change. What I have now should work, but it would be good if you take a closer look and suggest a better pattern to follow.
- The goroutine may return an error but there is no error channel for me to use to push the error upwards, so right now it just gets logged. Is there a better way to handle these errors?

azdagron commented:
Hey faisal, I appreciate your patience with the review here. We are heads down prepping for the 1.0.0 release this week. I'll make sure I get to this next week.

A bit of a nitpick, but if this is ready to go, can we remove WIP from the title?

faisal-memon commented:
> Hey faisal, I appreciate your patience with the review here. We are heads down prepping for the 1.0.0 release this week. I'll make sure I get to this next week.
> 
> A bit of a nitpick, but if this is ready to go, can we remove WIP from the title?

Thanks for checking in, np for the delay. Excited to see a 1.0.0 release is on the way. I've removed the WIP title.

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle_watcher.go:
```diff
@@ -0,0 +1,149 @@
+		if watcher != nil {
+			validWatcherPresent = true
+		}
+		watchers = append(watchers, watcher)
```
Seems a little funny and potentially dangerous to add a `nil` watcher to the list. Why not just only append for non-nil watchers? Then the `validWatcherPresent` bool can go away and a caller knows if there is valid watches simply by checking if `len(watchers) > 0`...

faisal-memon replied:
I removed the validWatcherPresent bool. [reflect.SelectCase](https://golang.org/pkg/reflect/#SelectCase) allows a chan to be nil, `If Chan is a zero Value, then the case is ignored...`.  In any case, I updated so the watcher is not even called if the `label` config is not set.

azdagron replied:
Even so, the presence of a `nil` watcher in the slice is fragile. Future maintenance could do something with that slice that would cause a panic if it ran into a nil watcher.

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -139,6 +144,22 @@ func (p *Plugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (resp
+	// Start watcher to set CA Bundle in objects created after server has started
+	watcher := newBundleWatcher(p)
+	go func() {
+		err := watcher.Start()
```
The bundle watcher usage pattern is a little weird. As you've pointed out there is nothing you can do if start fails but log, but that is because the initialization of the watchers and the actual watching is all lumped under start. What if newBundleWatcher did the initialization of the watchers  and returned an error on failure? Like so?
```
watcher, err := newBundleWatcher(p)
if err != nil {
    return nil, err
}
var wg sync.WaitGroup
ctx, cancel := context.WithBackground(context.Background()
wg.Add(1)
go func() {
    defer wg.Done()
    if err := watcher.Watch(ctx); err != nil && !error.Is(err, context.Canceled) {
        p.log.Error("Unable to watch: %v", err)
    |
}()

p.mu.Lock()
if p.cancelWatcher != nil {
    p.cancelWatcher()
}
p.cancelWatcher = func() {
    cancel()
    wg.Wait()
}
p.mu.Unlock()
```

Also, YMMV, but I've found the Start()/Stop() pattern  harder to test than a synchronous `Watch(ctx context.Context)` method.

faisal-memon replied:
Thanks for the suggestion, I wasn't happy with the start/stop pattern either.

faisal-memon replied:
This change is in

azdagron reviewed (commented):
Thank you for your patience, @faisal-memon. I've put together an initial set of comments here.

faisal-memon commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -233,24 +284,159 @@ func getKubeConfig(configPath string) (*rest.Config, error) {
+
+	// Step through all the the webhooks in the MutatingWebhookConfiguration
+	for i := range patch.Webhooks {
+		patch.Webhooks[i].AdmissionReviewVersions = mutatingWebhook.Webhooks[i].AdmissionReviewVersions
+		patch.Webhooks[i].ClientConfig.CABundle = []byte(bundleData(resp.Bundle))
+		patch.Webhooks[i].Name = mutatingWebhook.Webhooks[i].Name
+		patch.Webhooks[i].SideEffects = mutatingWebhook.Webhooks[i].SideEffects
```
Kubernetes throws an error if those fields are empty.

azdagron replied:
Huh, interesting.

faisal-memon commented on doc/plugin_server_notifier_k8sbundle.md:
```diff
@@ -14,6 +14,7 @@ The plugin accepts the following configuration options:
 | config_map            | The name of the ConfigMap                   | `spire-bundle`  |
 | config_map_key        | The key within the ConfigMap for the bundle | `bundle.crt`    |
 | kube_config_file_path | The path on disk to the kubeconfig containing configuration to enable interaction with the Kubernetes API server. If unset, it is assumed the notifier is in-cluster and in-cluster credentials will be used. | |
+| label                 | If set, rotate the CA Bundle in validating and mutating webhooks with this label set to `true`. | |
```
I had that originally, I moved it to regular `label` because there are other kubernetes types that have cabundles, for example `APIService` is one we use. I tried to write this to be extensible enough to add in these additional types as needed.

azdagron replied:
I don't know. I could see somebody being surprised to upgrade SPIRE and suddenly this plugin start watching and updating other resource types without operator buy-in.

@evan2645, curious to know your thoughts here.

faisal-memon replied:
We definitely don't want to surprise anyone. This new setting is disabled by default so if they use the same config when they upgrade, there won't be any change in behavior.

azdagron replied:
Sure. But I'm talking about a future change after this one where the plugin decides it wants to update additional K8S objects using the label. Operators who have already opted in to have the plugin manage webhooks would not have to explicitly opt in to get the new behavior for the new K8S objects. Furthermore, in order to _opt out_ of having those new objects managed, operators would have to define new labels, since the configurable is "all-or-nothing" for all object types with that label.

faisal-memon replied:
oh yea ok, that makes sense. cert-manager has them all lumped together so i was copying that pattern. We can have the types broken out into different configs here to avoid any surprises.

faisal-memon commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -161,22 +182,48 @@ func (p *Plugin) setConfig(config *pluginConfig) {
+				return err
+			}
+			if err := p.updateBundle(ctx, c, client, itemMeta.GetNamespace(), itemMeta.GetName()); err != nil {
+				return k8sErr.New("unable to update %s/%s: %v", itemMeta.GetNamespace(), itemMeta.GetName(), err)
```
I made this change. itll update all the bundles and cumulate the errors. The SPIRE Server will still crash if there are any errors updating bundles. Do you think this is correct behavior going forward?

azdagron replied:
The crash only happens on the initial  BundleLoaded event. Errors from BundleUpdated events are ignored and logged. I think in these circumstances, best effort updating is appropriate.

faisal-memon commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -139,6 +144,22 @@ func (p *Plugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (resp
 
 	p.setConfig(config)
+
+	// Start watcher to set CA Bundle in objects created after server has started
```
Moved to setConfig.

faisal-memon commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle_watcher.go:
```diff
@@ -0,0 +1,149 @@
+		b.mu.Unlock()
+	}()
+
+	config, err := b.p.getConfig()
```
done

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -138,8 +144,7 @@ func (p *Plugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (resp
 
-	p.setConfig(config)
-	return &spi.ConfigureResponse{}, nil
+	return &spi.ConfigureResponse{}, p.setConfig(config)
```
nit: even though gRPC will handle it properly, we shouldn't return a non-nil response with an error. I'd suggest splitting this up and returning an explicit `nil, err` if `setConfig` fails.

faisal-memon replied:
done

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -155,28 +160,84 @@ func (p *Plugin) getConfig() (*pluginConfig, error) {
 	p.config = config
+
+	// Start watcher to set CA Bundle in objects created after server has started
+	if p.cancelWatcher != nil {
+		p.cancelWatcher()
+	}
```
The order of operations here could be a little more robust. Currently, if there is a failure creating the watcher, 1) the `Configure` call will return an error but will have already mutated the configuration (i.e. `p.config = config`) and 2) the original watcher will have already been shut down. I suggest the following order of events:

- if label configured
    - create new watcher
    - cancel existing watcher and replace `cancelWatcher`
- replace config

faisal-memon replied:
that makes sense. Only change i would like is to cancel existing watcher outside of ` if label is configured` for the scenario when the label is configures and then the config is updated and the label is no longer configured. We still need to cancel the watcher in that scenario.

azdagron replied:
Yep, good catch.

faisal-memon replied:
this change is in

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -155,28 +160,84 @@ func (p *Plugin) getConfig() (*pluginConfig, error) {
+	for _, client := range clients {
+		list, err := client.GetList(ctx, c)
+		if err != nil {
+			return k8sErr.New("unable to get list: %v", err)
```
should this error also just be aggregated?

faisal-memon replied:
all these errors are now aggregated.

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -155,28 +160,84 @@ func (p *Plugin) getConfig() (*pluginConfig, error) {
+		}
+		listItems, err := meta.ExtractList(list)
+		if err != nil {
+			return k8sErr.New("unable to extract list items: %v", err)
```
should this error also just be aggregated?

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle.go:
```diff
@@ -155,28 +160,84 @@ func (p *Plugin) getConfig() (*pluginConfig, error) {
+		for _, item := range listItems {
+			itemMeta, err := meta.Accessor(item)
+			if err != nil {
+				return err
```
should this error also just be aggregated?

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle_watcher.go:
```diff
@@ -0,0 +1,111 @@
+func (b *bundleWatcher) newWatchers(ctx context.Context) ([]watch.Interface, error) {
+	watchers := []watch.Interface{}
+	for _, client := range b.clients {
+		watcher, err := client.Watch(ctx, b.config.Label)
```
As far as I can tell, client.Watch will still be invoked on the configMapClient, producing a `nil` watcher interface. We should filter those out and not add them to the `watchers` slice, even if `newSelectCase` handles it. Future maintenance could change the interactions with the watchers slice, causing panics. A `nil` interface value in a slice is a little smelly and should be avoided.

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle_watcher.go:
```diff
@@ -0,0 +1,111 @@
+				Dir:  reflect.SelectRecv,
+				Chan: reflect.ValueOf(watcher.ResultChan()),
+			})
+		} else {
```
As mentioned previously, we should refactor such that a `nil` watcher can't end up in the slice. This condition should be removed.

azdagron commented:
@faisal-memon, please let me know when you are ready for another review pass on this.

faisal-memon commented:
> @faisal-memon, please let me know when you are ready for another review pass on this.

Hey @azdagron, ready for another review pass, ive addressed all the comments.

azdagron reviewed (commented):
Things are looking pretty good here.

I went ahead and gave this a go. One thing I ran into is that when naively reapplying my yaml for the webhook, I accidentally replaced the caBundle value that was added by this plugin. I'm wondering if we shouldn't respond to `updated` events by updating the caBundle value if-and-only-if it is empty. This allows an operator to overwrite the value with something else, but if the value gets removed, the plugin will restore it....

Thoughts?

faisal-memon commented:
> Things are looking pretty good here.
> 
> I went ahead and gave this a go. One thing I ran into is that when naively reapplying my yaml for the webhook, I accidentally replaced the caBundle value that was added by this plugin. I'm wondering if we shouldn't respond to `updated` events by updating the caBundle value if-and-only-if it is empty. This allows an operator to overwrite the value with something else, but if the value gets removed, the plugin will restore it....
> 
> Thoughts?

Ooo interesting edge case. I think we should handle that case for sure. It doesn't make sense to me for the operator to fill it with something else though because when it comes time to rotate the cabundle itll get overwritten by SPIRE anyways. To me if you set the label, you are handing that field over to SPIRE to manage.

azdagron commented:
> if you set the label, you are handing that field over to SPIRE to manage.

That's probably a fair assumption. One thought, can the plugin easily filter out update events caused by itself?

faisal-memon commented:
> > if you set the label, you are handing that field over to SPIRE to manage.
> 
> That's probably a fair assumption. One thought, can the plugin easily filter out update events caused by itself?

Yea thats a good point. There's not built in way that I know of. We'd have to check if the caBundle matches the current root cert and drop the event.

faisal-memon commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle_watcher.go:
```diff
@@ -0,0 +1,112 @@
+		chosen, recv, _ := reflect.Select(selectCase)
+		if chosen < len(b.clients) {
+			if err := b.watchEvent(ctx, b.clients[chosen], recv.Interface().(watch.Event)); err != nil {
+				return k8sErr.New("handling watch event: %v", err)
```
I ran into a boostrapping issue when testing today, gettting this error here: 
```
time="2021-03-09T23:14:44Z" level=error msg="Unable to watch" error="k8s-bundle: handling watch event: rpc error: code = FailedPrecondition desc = IdentityProvider host service has not been initialized" subsystem_name=built-in_plugin.k8sbundle
```
It happens when  a webhook is deployed prior to the spire server being deployed. The watcher gets set up before the IdentityProvider gets setup, and we get the add event from the webhook. Is there a place we can hook into to set up the watcher after IdentityProvider is initialized?

azdagron replied:
Giving it some thought, option 1 is probably easiest considering all of the mechanics in Configure() around watcher reconfiguration, etc.

faisal-memon replied:
Ok i made this change. I also made the watcher not exit on error to be more consistent with the way updateBundles behaves. LMK if this ok or if you prefer it to exit on error.

azdagron commented on pkg/server/plugin/notifier/k8sbundle/k8sbundle_watcher.go:
```diff
@@ -62,17 +64,21 @@ func (b *bundleWatcher) watch(ctx context.Context) error {
+			// to populate the caBundle, so its safe to ignore this error.
 			return err
 		}
+		b.p.log.Debug("Set bundle for object", "name", objectMeta.GetName(), "event", event.Type)
```
This requires a slight reordering otherwise we'll log that we set the bundle when we encounter FailedPrecondition, maybe something like:
```
err = b.p.updateBundle(ctx, b.config, client, objectMeta.GetNamespace(), objectMeta.GetName())
switch {
case err == nil:
	b.p.log.Debug("Set bundle for object", "name", objectMeta.GetName(), "event", event.Type)
case status.Code(err) == codes.FailedPrecondition:
	// Ignore FailPrecondition errors for when SPIRE is booting and we receive an event prior to
	// IdentityProvider being initialized. In this case the BundleLoaded event will come
	// to populate the caBundle, so its safe to ignore this error.
default:
    return err
}```

azdagron commented:
> We'd have to check if the caBundle matches the current root cert and drop the event.
I'm trying to think through what that means for HA and multiple servers receiving these update events and I _think_ this will settle ok...

faisal-memon commented:
> > We'd have to check if the caBundle matches the current root cert and drop the event.
> 
> I'm trying to think through what that means for HA and multiple servers receiving these update events and I _think_ this will settle ok...

Thats good. This change is now in.

faisal-memon commented:
> > > We'd have to check if the caBundle matches the current root cert and drop the event.
> > 
> > 
> > I'm trying to think through what that means for HA and multiple servers receiving these update events and I _think_ this will settle ok...
> 
> Thats good. This change is now in.

Did some testing today with 3 spire servers and everything goes through ok. Only one server processes each event.

azdagron reviewed (approved):
\o/ awesome, i think we're good here. Thanks for seeing this through, @faisal-memon !
